File: individual.py
import numpy as np
import rnn
from load_config import load_config

config = load_config()

# morphology
EYES = config["eyes"]

# rnn layers
INPUT_LAYER_SIZE = EYES
HIDDEN_LAYERS = config["hidden_layers"]
HIDDEN_LAYERS_SIZE = config["hidden_layers_size"]
OUTPUT_LAYER_SIZE = config["output_layer_size"]
# number of chromosome pairs
CHROMOSOME_LENGTH = HIDDEN_LAYERS_SIZE
APLOID_NUMBER = INPUT_LAYER_SIZE + HIDDEN_LAYERS_SIZE * (2 * HIDDEN_LAYERS - 1) + OUTPUT_LAYER_SIZE

# constants
TURN = 2 * np.pi
FIELD_OF_VIEW = TURN / (2 * EYES)
EYE_ANGLES = (2 * np.arange(EYES).reshape((EYES, 1)) + 1) * FIELD_OF_VIEW / 2
LIGHT_INTENSITY = config["light_intensity"]

DEG = 180 / np.pi

class individual:
	@staticmethod
	def create():
		# generates APLOID_NUMBER random chromosomes
		return individual(np.random.randn(APLOID_NUMBER, 1, CHROMOSOME_LENGTH), np.random.randn(APLOID_NUMBER, 1, CHROMOSOME_LENGTH))

	def __init__(self, f_chrom, m_chrom):
		self.fitness = 0
		# stores the chromosomes
		self.f_chrom = f_chrom
		self.m_chrom = m_chrom
		# rebuild the weights from genes
		fusion = (self.f_chrom + self.m_chrom) / 2
		# split indexes
		recurrences = HIDDEN_LAYERS_SIZE * HIDDEN_LAYERS
		inputs = recurrences + INPUT_LAYER_SIZE
		hiddens = inputs + recurrences - HIDDEN_LAYERS_SIZE
		self.brain = rnn.RNN(
			# recurrences
			fusion[:recurrences, 0, :].reshape(HIDDEN_LAYERS, HIDDEN_LAYERS_SIZE, HIDDEN_LAYERS_SIZE),
			# input -> hiddens
			[fusion[recurrences:inputs, 0, :].T] +
			# hiddens -> hiddens
			map(np.squeeze, np.vsplit(fusion[inputs:hiddens, 0, :], HIDDEN_LAYERS - 1)) +
			# hiddens -> output
			[fusion[hiddens:, 0, :]]
		)
		self.reset()

	def reset(self):
		self.position = np.zeros((1, 2))
		self.rotation = TURN / 8
		self.brain.reset()

	def tick(self, input):
		# outputs are changes to position and orientation
		[[delta_x], [delta_y], [delta_theta]] = self.brain.forward(input)
		# rotates displacement
		sin_rot = np.sin(self.rotation)
		cos_rot = np.cos(self.rotation)
		self.position += np.dot([[delta_x, delta_y]], [[cos_rot, sin_rot], [-sin_rot, cos_rot]])
		self.rotation = (self.rotation + delta_theta * TURN) % TURN

	def visibility(self, point):
		(x, y) = point
		# find inclination of the point (angle of the line between origin and point)
		y_dist = y - self.position[0, 1]
		x_dist = x - self.position[0, 0]
		angle = np.arctan2(y_dist, x_dist)
		# intensity dimishes with the square of the distance... or so
		d_squared = y_dist ** 2 + x_dist ** 2
		intensity = LIGHT_INTENSITY / (d_squared / 2)
		# max and min angle from which you can see 'point'
		theta_min = np.arcsin(1 / np.sqrt(d_squared))
		theta_max = TURN / 4 - 2 * theta_min
		# rotate wrt the point inclination
		theta_min += angle - TURN / 4
		theta_max += angle
		# get the angle of incidence of light at angle phi
		get_theta = lambda phi: np.abs(np.arctan2(x - np.sin(phi), y - np.cos(phi)) + phi)
		# rotate the eyes wrt animal orientation
		phis = EYE_ANGLES - self.rotation
		# compute incidence on the eyes
		return (intensity * np.sin(get_theta(phis))) * (phis > theta_min) * (phis < theta_max)

	def gamete(self):
		gamete = np.empty((APLOID_NUMBER, 1, CHROMOSOME_LENGTH))
		for i in xrange(APLOID_NUMBER):
			# toss a coin to choose wheter use a chromosome
			# from the father genes or the mother's as the main one
			m_or_f = np.random.rand() < 0.5

			# sometimes crossing over happens
			r = np.random.rand()
			if r < config["crossing_over_rate"]:
				# draw the locus where to split
				r = np.random.randint(CHROMOSOME_LENGTH)
				gamete[i, 0, 0:r] = (self.f_chrom if m_or_f else self.m_chrom)[i, 0, :r]
				gamete[i, 0, r:] = (self.m_chrom if not m_or_f else self.f_chrom)[i, 0, r:]
			else:
				r = np.random.rand()
				gamete[i] = (self.f_chrom if m_or_f else self.m_chrom)[i, 0]

			# sometimes, mutate a gene
			r = np.random.rand()
			if r < config["mutation_rate"]:
				# draw the gene to mutate
				r = np.random.randint(CHROMOSOME_LENGTH)
				gamete[i, 0, r] = np.random.randn() * gamete[i, 0, r]

		return gamete

	def mate(self, partner):
		return individual(self.gamete(), partner.gamete())

File: cpp/individual.cpp
#include <iostream>
#include <cmath>

#ifndef M_2_PI
	#define M_2_PI 6.2832
#endif
#ifndef M_PI_2
	#define M_PI_2 1.5708
#endif
#ifndef M_PI_4
	#define M_PI_4 0.78540
#endif

// morphology
#define EYES 5

// neural network layers size
#define INPUTS EYES
#define HIDDEN_SIZE 6
#define HIDDEN_LAYERS 2
#define OUTPUTS 3

// number of chromosome pairs
#define GENOME_SIZE (HIDDEN_SIZE * (INPUTS + 1) + (HIDDEN_LAYERS - 1) * (HIDDEN_SIZE * (HIDDEN_SIZE + 1)) + OUTPUTS * (HIDDEN_SIZE + 1))
// probability of mutation events
#define CROSSING_OVER_RATE 0.1
#define MUTATION_RATE 0.2

// weights interval
#define INTERVAL 1.0

#include "rnn.cpp"

// constants
#define LIGHT_INTENSITY 1024.0

struct gene {
	double value;
	bool dominant;
};

bool cache_eye_angles = false;
double EYE_ANGLES[EYES];
int CHROMOSOME_LENGTHS[] = {INPUTS + 1, HIDDEN_SIZE + 1, OUTPUTS};
int CHROMOSOME_NUMBER[] = {HIDDEN_SIZE, HIDDEN_SIZE, HIDDEN_SIZE + 1};

class Individual {
	public:
		double fitness = 0.0;
		double rotation;
		double position[2];
		NeuralNetwork* brain;
		gene* mother_genes = nullptr;
		gene* father_genes = nullptr;
		double* expressed_genes;

		static void gen_angles() {
			// generate EYE_ANGLES if not done before
			if (!cache_eye_angles) {
				for (int i = 0; i < EYES; ++i) {
					EYE_ANGLES[i] = (2 * i + 1) * (M_PI_2 / (double)EYES);
				}
				cache_eye_angles = true;
			}
		}

		static Individual* create() {
			gen_angles();
			// creates two random gametes
			gene* gametes[2];
			gametes[0] = new gene[GENOME_SIZE];
			gametes[1] = new gene[GENOME_SIZE];

			for (int i = 0; i < 2; ++i) {
				for (int g = 0; g < GENOME_SIZE; ++g) {
					gametes[i][g].value = fuzzy_rand();
					gametes[i][g].dominant = fuzzy_rand() < 0.0;
				}
			}
			return new Individual(gametes[0], gametes[1]);
		}

		Individual(gene* gamete1, gene* gamete2) {
			this->mother_genes = gamete1;
			this->father_genes = gamete2;
			double* genome = new double[GENOME_SIZE];
			// simulates dominance/recessivity
			#pragma ivdep
			for (int i = 0; i < GENOME_SIZE; ++i) {
				genome[i] =  this->mother_genes[i].dominant ? this->mother_genes[i].value : this->father_genes[i].value;
				genome[i] += this->father_genes[i].dominant ? this->father_genes[i].value : this->mother_genes[i].value;
				genome[i] *= INTERVAL / 2.0;
			}
			setup(genome);
		}
		Individual(double* genome) {
			gen_angles();
			setup(genome);
		}
		void setup(double* genome) {
			this->expressed_genes = genome;
			this->brain = new NeuralNetwork(this->expressed_genes);
			this->reset();
		}

		void reset() {
			this->position[0] = this->position[1] = 0.0;
			this->rotation = M_PI_4;
		}

		void tick(double* food) {
			// outputs are changes to position and orientation
			double delta_x, delta_y, delta_rot;
			auto output = this->brain->forward(this->visibility(food));
			delta_x = output(0);
			delta_y = output(1);
			delta_rot = output(2);
			// rotates displacement
			double sin_rot = sin(this->rotation);
			double cos_rot = cos(this->rotation);
			this->position[0] += delta_x *  cos_rot + delta_y * sin_rot;
			this->position[1] += delta_x * -sin_rot + delta_y * cos_rot;
			this->rotation = fmod(this->rotation + delta_rot * M_2_PI, M_2_PI);
		}

		input_vector_t visibility(double* food) {
			// find inclination of the point (angle of the line between origin and point)
			double x_dist = food[0] - this->position[0];
			double y_dist = food[1] - this->position[1];
			double angle = atan2(y_dist, x_dist);
			// intensity dimishes with the square of the distance... or so
			double d_squared = pow(y_dist, 2) + pow(x_dist, 2);
			double intensity = 2.0 * LIGHT_INTENSITY / d_squared;
			// max and min angle from which you can see 'food'
			double theta_min = asin(1.0 / sqrt(d_squared));
			double theta_max = M_PI_2 - 2.0 * theta_min;
			// rotate wrt the point inclination
			theta_min += angle - M_PI_2;
			theta_max += angle;

			// compute visibility of food for every eye
			double phi[EYES];
			input_vector_t visibility;
			double *raw = &(visibility(0));

			for (int i = 0; i < EYES; ++i) {
				phi[i] = EYE_ANGLES[i] - this->rotation;
			}

			#pragma ivdep
			for (int i = 0; i < EYES; ++i) {
				raw[i] = (theta_min < phi[i]) * (phi[i] < theta_max) * intensity * sin(fabs(atan2(food[1] - sin(phi[i]), food[0] - cos(phi[i])) + phi[i]));
			}

			return visibility;
		}

		void chr_mitosis(gene* gamete, int offset, int chr_length) {
			bool m_or_f;
			int split_loc, mutation_loc;
			// alias for parents genes
			gene* m = this->mother_genes;
			gene* f = this->father_genes;

			// shift to offset
			gamete += offset;

			// toss a coin to choose wheter use a chromosome
			// from the father genes or the mother's as the "main" one
			m_or_f = fuzzy_rand() < 0.0;

			// sometimes, crossing over happens
			if (fuzzy_rand() < (CROSSING_OVER_RATE * 2 - 1)) {
				// draw the locus where to split
				split_loc = floor(chr_length * (fuzzy_rand() + 1) / 2);
				std::memcpy(gamete, (m_or_f ? m : f) + offset, split_loc * sizeof(gene));
				std::memcpy(gamete + split_loc, (m_or_f ? f : m) + offset + split_loc, (chr_length - split_loc) * sizeof(gene));
			} else {
				std::memcpy(gamete, (m_or_f ? m : f) + offset, chr_length * sizeof(gene));
			}

			// sometimes, mutate a gene
			if (fuzzy_rand() < (MUTATION_RATE * 2 - 1)) {
				// mutate a randomly drawn gene
				mutation_loc = floor(chr_length * (fuzzy_rand() + 1) / 2);
				gamete[mutation_loc].value = fuzzy_rand();
				gamete[mutation_loc].dominant = fuzzy_rand() < 0.0;
			}
		}

		gene* gamete() {
			// new genes
			gene *gamete = new gene[GENOME_SIZE];
			int offset = 0;

			for (int c = 0; c < 3; ++c) {
				for (int i = 0; i < CHROMOSOME_NUMBER[c]; i += CHROMOSOME_LENGTHS[c]) {
					this->chr_mitosis(gamete, offset, CHROMOSOME_LENGTHS[c]);
					offset += CHROMOSOME_LENGTHS[c];
				}
			}
			return gamete;
		}

		Individual* mate(Individual* partner) {
			return new Individual(this->gamete(), partner->gamete());
		}

		~Individual() {
			// free allocated memory
			if (this->father_genes != nullptr) delete [] this->father_genes;
			if (this->mother_genes != nullptr) delete [] this->mother_genes;
			delete [] this->expressed_genes;
			delete this->brain;
		}
};
